vsintegration/src/FSharp.Editor/Common/LanguageService.fs

```diff
@@ -245,28 +242,38 @@ and
     internal FSharpLanguageService(package : FSharpPackage) =
     inherit AbstractLanguageService<FSharpPackage, FSharpLanguageService>(package)
 
-    let checkerProvider = package.ComponentModel.DefaultExportProvider.GetExport<FSharpCheckerProvider>().Value
     let projectInfoManager = package.ComponentModel.DefaultExportProvider.GetExport<ProjectInfoManager>().Value
 
     let projectDisplayNameOf projectFileName = 
         if String.IsNullOrWhiteSpace projectFileName then projectFileName
         else Path.GetFileNameWithoutExtension projectFileName
 
-    let openedProjects = Queue<IProvideProjectSite>()
+    let singleFileProjects = ConcurrentDictionary<_, AbstractProject>()
 
-    do
-      Events.SolutionEvents.OnAfterOpenProject.Add (fun (args: Events.OpenProjectEventArgs) -> 
-        match args.Hierarchy with
-        | :? IProvideProjectSite as siteProvider ->
-            openedProjects.Enqueue(siteProvider)            
-        | _ -> ())
-        
-    member this.BatchSetupProjects() =    
-        if openedProjects.Count > 0 then    
-            let workspace = package.ComponentModel.GetService<VisualStudioWorkspaceImpl>()
-            for siteProvider in openedProjects do
-                this.SetupProjectFile(siteProvider, workspace)
-            openedProjects.Clear()
+    let tryRemoveSingleFileProject projectId =
+        match singleFileProjects.TryRemove(projectId) with
+        | true, project ->
+            projectInfoManager.RemoveSingleFileProject(projectId)
+            project.Disconnect()       
+        | _ -> ()
+
+    override this.Initialize() =
+        base.Initialize()
+ 
+        this.Workspace.Options <- this.Workspace.Options.WithChangedOption(Completion.CompletionOptions.BlockForCompletionItems, FSharpCommonConstants.FSharpLanguageName, false)
+        this.Workspace.Options <- this.Workspace.Options.WithChangedOption(Shared.Options.ServiceFeatureOnOffOptions.ClosedFileDiagnostic, FSharpCommonConstants.FSharpLanguageName, Nullable false)
+
+        this.Workspace.DocumentClosed.Add <| fun args ->
+            tryRemoveSingleFileProject args.Document.Project.Id
+        //because DocumentClosed doesn't fire when closing solution, we need also this:
+        Events.SolutionEvents.OnAfterCloseSolution.Add <| fun _ ->
+            for id in singleFileProjects.Keys do tryRemoveSingleFileProject id |> ignore
+            
+        Events.SolutionEvents.OnAfterOpenProject.Add <| fun args ->
```

Comment:
I think there is some potential to speed things up. Here's a rough draft that should bring the solution load speed up to the current level [https://github.com/majocha/visualfsharp/tree/lsmailbox](https://github.com/majocha/visualfsharp/tree/lsmailbox) together with this PR